src/span.js

```diff
@@ -0,0 +1,163 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import assert from 'assert';
+import * as thrift from './thrift.js';
+import Utils from './util.js';
+import SpanContext from './span_context.js';
+
+export default class Span {
+    _tracer: any;
+    _name: string;
+    _spanContext: SpanContext;
+    _start: number;
+    _duration: number;
+    _port: number;
+    _isClient: boolean;
+    _peer: Endpoint;
+    _ended: boolean;
+    _annotations: Array<Annotation>;
+    _binaryAnnotations: Array<BinaryAnnotation>;
+
+    constructor(tracer: any,
+                name: string,
+                spanContext: SpanContext,
+                start: number,
+                isClient: boolean = false,
+                peer: Endpoint,
+                annotations: Array<Annotation> = [],
+                binaryAnnotations: Array<BinaryAnnotation> = []
+    ) {
+        this._tracer = tracer;
+        this._name = name;
+        this._spanContext = spanContext;
+        this._start = start;
+        this._duration = 0;
+        this._port = 0;
+        this._isClient = isClient;
+        this._peer = peer;
+        this._ended = false;
+
+        // logs
+        this._annotations = annotations;
+
+        // tags
+        this._binaryAnnotations = binaryAnnotations;
+    }
+
+    /**
+     * Returns the span context that represents this span.
+     *
+     * @return {SpanContext} - Returns this span's span context.
+     **/
+    context(): SpanContext {
+        return this._spanContext;
+    }
+
+    /**
+     * Returns the tracer associated with this span.
+     * @return {Tracer} - returns the tracer associated witht this span.
+     **/
+    tracer(): any {
+        return this._tracer;
+    }
+
+    /**
+     * Sets the operation name on this given span.
+     *
+     * @param {string} name - The name to use for setting a span's operation name.
+     **/
+    setOperationName(name: string): void {
+        this._name = name;
+    }
+
+    /**
+     * Finishes a span which has the effect of reporting it and
+     * setting the finishTime on the span.
+     *
+     * @param {number} finishTime - The time on which this span finished.
+     **/
+    finish(finishTime: number): void {
+        assert(!this._ended, "You can only call finish() on span once");
+        this._ended = true;
+
+        if (this._spanContext.IsSampled() && this._start) {
+            let endTime = finishTime || Utils.getTimestampMicros();
+            this._duration = endTime - this._start;
+            this.log({
+                event: this._isClient ? thrift.CLIENT_RECV : thrift.SERVER_SEND,
+                timestamp: endTime
+            });
```

Comment:
I think we should move these to reporter. They are artifacts of Zipkin representation, we don't need them to be in the main tracer code.